### PR TITLE
Editor: Allow title to split onto multiple lines

### LINKED ISSUE #1059: Editor: allow long titles to wrap to multiple lines
Raised by @nylen 

Allow long post/page titles to wrap to multiple lines.  After 9892-gh-calypso-pre-oss here is the current behavior:

title-fade-unless-editing

We want the title to wrap onto multiple lines if needed:

image

Depends on #337 being completed first.

### LINKED ISSUE #1679: Editor: Mobile: cannot scroll within long titles in Safari on iOS
Raised by @rachelmcr 

What I Did:

1. Created a new post and added a long title (more than one screen width):

2. Tried to edit the end of the post title

What I Expected:

To be able to tap and drag the cursor to the end of the title, so I could edit/erase it.

What Actually Happened:

When I tapped and dragged the cursor, it didn't scroll past the part of the title that appeared on screen. Instead, the cursor moved all the way to the right edge of the screen:

Device: iPhone 6, iOS 9.0.2, Safari

Notes from @kriskarkoski 

> Just noticed something similar while testing, where I couldn't scroll back to the beginning of the title and clicking into the content area didn't reset it to the beginning right away, but rather after a little time passed
> 
> Also, I can scroll on Chrome on iOS 9, but it feels less instant than it should be, instead accelerating the scroll rate after I hold it to one edge for a moment rather than scrolling instantly

Related: #1059

### PR DESCRIPTION
**Pull request by aduth**
~~Blocked by #4824~~
Closes #1059
Fixes #1679

This pull request seeks to allow the editor post title to split onto multiple lines, rather than obscuring it with a long-content fade when the input loses focus.

It also introduces a new `<TextareaAutosize />` component to serve as a drop-in replacement for `<textarea />` in order to facilitate in this use-case.

Before|After
---|---
before|after

__Implementation notes:__

While the implementation does use [`autosize`](https://github.com/jackmoore/autosize), a third-party library, note that this is already used in the editor for the HTML editing tab, so the change in editor bundle size should be negligible if anything. A follow-up task may be to consider whether the HTML tab textarea can be updated to use this component instead.

__Testing instructions:__

Verify in a large and small viewport size that the editor title resizes to accommodate multiple lines as needed. Ensure that existing behavior remains unaffected (notably, focusing the input for a new post, including when navigating from an existing post to a new post).

1. Navigate to the [Reader](http://calypso.localhost:3000/post)
2. Start a new post using the New Post masterbar button
3. Note that the title field is auto-focused for your new post
4. Enter text into the title area
5. Note that the textarea expands to allow multiple line entry

### CODE REVIEW

File: client/post-editor/editor-title/index.jsx [hunk 1]
```diff
@@ -118,17 +97,17 @@ const EditorTitle = React.createClass( {
...
 						placeholder={ this.translate( 'Title' ) }
 						onChange={ this.onChange }
 						onBlur={ this.onBlur }
-						onFocus={ this.onFocus }
 						autoFocus={ isNew && ! isMobile() }
 						value={ post ? post.title : '' }
 						aria-label={ this.translate( 'Edit title' ) }
-						ref="titleInput" />
+						ref="titleInput"
+						rows="1" />
```
Comment (mtias) [inline client/post-editor/editor-title/index.jsx]: I assume this is the default?

**Comment by mtias** [comment]
This is great, thanks for looking at it after I complained :)

It works well for me, and I wouldn't mind merging before #4824 if that may take longer.

File: client/post-editor/editor-title/style.scss [hunk 2]
```diff
@@ -1,10 +1,6 @@
 .editor-title {
```
Comment (mtias) [inline client/post-editor/editor-title/style.scss]: Is there anything we could simplify markup-wise here? Some of these nested elements were needed for correctly rendering the fade effect, which we don't need anymore.

File: client/post-editor/editor-title/index.jsx (hunk 1 above)
Comment (aduth) [inline client/post-editor/editor-title/index.jsx]: Nope, the `rows` default is `2` ([reference](https://docs.webplatform.org/wiki/html/elements/textarea))

File: client/post-editor/editor-title/style.scss (hunk 2 above)
Comment (aduth) [inline client/post-editor/editor-title/style.scss]: >Is there anything we could simplify markup-wise here?

We'll likely need the relative wrapper for the loading block still, but I'll make another pass to be certain.

File: client/post-editor/editor-title/index.jsx (hunk 1 above)
Comment (mtias) [inline client/post-editor/editor-title/index.jsx]: Right, I mean the default initial value for the component.

File: client/post-editor/editor-title/style.scss (hunk 2 above)
Comment (mtias) [inline client/post-editor/editor-title/style.scss]: We should remove the loading block :)

Comment (aduth) [inline client/post-editor/editor-title/style.scss]: >We should remove the loading block :)

One PR at a time 😄

**Comment by aduth** [comment]
>It works well for me, and I wouldn't mind merging before #4824 if that may take longer.

I want to do some last-minute performance debugging, as I was noticing an input lag in my testing. I think it may be that both the `value` assignment of the input field and the input's own `change` event (bound by `autosize`) are duplicating effort in resizing the field.

File: client/post-editor/editor-title/index.jsx (hunk 1 above)
Comment (aduth) [inline client/post-editor/editor-title/index.jsx]: >Right, I mean the default initial value for the component.

Still not certain I'm following, but an empty field will default to `1` rows, and grow as necessary via `autosize`. Or are you suggesting we use `rows="1"` as a default prop for `<TextareaAutosize />` ?

Comment (mtias) [inline client/post-editor/editor-title/index.jsx]: I didn't know what the prop meant just by looking at it :)

**Comment by aduth** [comment]
This was not previously accounting for newlines, added either by enter keypress or pasted text including multiple lines. Both cases are now accounted for in 99d89b1. Flipping this back to Needs Review.

File: client/post-editor/editor-title/index.jsx [hunk 3]
```diff
@@ -45,37 +41,40 @@ const EditorTitle = React.createClass( {
...
-			( nextProps.isNew && ( this.props.site && nextProps.site ) && ( this.props.site.ID !== nextProps.site.ID ) )
-		) {
-			this.setState( {
-				isFocused: true
-			}, () => {
-				this.refs.titleInput.focus();
-			} );
+		const { isNew, site } = this.props;
+		if ( ( isNew && ! prevProps.isNew ) ||
+				( isNew && get( prevProps.site, 'ID' ) !== get( site, 'ID' ) ) ) {
```
Comment (nylen) [inline client/post-editor/editor-title/index.jsx]: I think this indentation style is clearer here:

```js
if (
    ( isNew && ! prevProps.isNew ) ||
    ( isNew && get( prevProps.site, 'ID' ) !== get( site, 'ID' ) )
) {
    // ...
}
```

Comment (aduth) [inline client/post-editor/editor-title/index.jsx]: >I think this indentation style is clearer here:

I go back and forth myself, but FWIW this is the documented style:

https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#multi-line-statements

File: client/post-editor/editor-title/index.jsx [hunk 4]
```diff
@@ -85,27 +84,48 @@ const EditorTitle = React.createClass( {
...
+		text = text.replace( /\s+/g, ' ' ).trim();
+
+		// Splice trimmed text into current title selection
+		const { value, selectionStart, selectionEnd } = event.target;
+		const valueChars = value.split( '' );
+		valueChars.splice( selectionStart, selectionEnd - selectionStart, text );
+		const newTitle = valueChars.join( '' );
+
+		this.setTitle( newTitle );
+		this.selectionStart = selectionStart + text.length;
```
Comment (nylen) [inline client/post-editor/editor-title/index.jsx]: Could use an explanatory comment:

```js
// Cursor position may need to change, but we have to wait for the value to update first
```

**Comment by nylen** [comment]
Works on Chrome, Safari, Firefox, Edge.  Code looks good.

There are a couple of minor issues on IE11:

long-title-ie11

- Scrollbar (apparently [can be eliminated with `overflow: auto`](https://css-tricks.com/snippets/css/remove-scrollbar-from-textarea-in-ie/))
- Cursor resets to end of line on paste.  (Attempting to debug this crashed IE in my VM.)

On iOS, I'm not able to paste because the text toolbar disappears immediately (also an issue in master - I will make a new issue), but other than that, the multi-line behavior seems fine.

**Comment by aduth** [comment]
@nylen : Thanks for the review. Both issues should be fixed now. Neither solutions are pretty 😄  (692ecda, cab2d16).

I'm still not entirely happy with the performance of the input with these changes. I've confirmed that the update calculations aren't being performed multiple times, but it still doesn't feel quite right. I think it's likely a combination of both the height calculation and the title saving its changes directly to the post store on each change. I'd like to move away from the latter case, but I think there are UX implications of having the option to "Save" changes as soon as characters are entered into the title field.

**Comment by nylen** [comment]
👍 :shipit:

**Comment by lancewillett** [comment]
OK to merge? Waiting on anything? (5 days ago.)

**Comment by aduth** [comment]
@lancewillett Works but not very happy with performance. Would be curious to get more feedback on whether it's noticeable, or at least spend more time profiling / improving performance before merging (as mentioned above, may require parallel changes to how we're tracking post title state).

**Comment by nylen** [comment]
fwiw, I didn't notice any performance problems specific to this change.  The editor was pretty laggy in my IE11 VM, but that was a general issue and also true in master.

**Comment by lancewillett** [comment]
I'm marking as "In Progress" so it doesn't look like it's ready to merge by anyone.

**Comment by aduth** [comment]
Updated original comment to reflect that this incidentally fixes #1679.

**Comment by lancewillett** [comment]
Hi @aduth — checking in on this one since last update was about a month ago. Still in progress? Can we help test a branch or change to nudge it ahead?

**Comment by aduth** [comment]
@lancewillett : I'm still interested in this one. I proposed #5445 in hopes of potentially addressing performance concerns, but may still want to merge this one beforehand nonetheless. At the least, I'll plan to rebase and do a last round of testing, as some of the issues I may have been experiencing could have been exclusive to my IE11 VM (where issues exist in master as well).

**Comment by aduth** [comment]
Rebased to resolve merge conflicts. @lancewillett @mtias @nylen one of you want to take this for a final spin? Else, since it'd previously been marked as ready to merge, I may move forward with merging in the next day.

I tried it again in IE11 and, even with my slow VM, it wasn't noticeable enough an issue that I'd consider performance a blocker for merging (may have improved since merging #5438).

**Comment by nylen** [comment]
Upon retesting, I'm seeing some issues with this.  I think they are a mix of existing issues that I missed before (pasting, undo stack) and new ones (Safari resize behavior).  In the recordings below, I'm typing, then cut/pasting some text, and then hitting undo/redo.

## Chrome

In Chrome (v50), pasting should not trim the initial space, and pasting also makes undo/redo act in unexpected ways:

editor-multiline-titles-chrome

## Safari

In Safari (v9.1.1), there are multiple issues:

- pasting trims the inital space like in Chrome
- the undo stack is differently broken (only the paste event is retained, which I guess is not as bad)
- the elastic scrolling effect briefly interferes with the resizing
- the textarea does not decrease its height when it should:

editor-multiline-titles-safari

## Firefox

In Firefox, everything works really well.

**Comment by aduth** [comment]
Thanks @nylen , these are some good catches. With regards to the pasting oddities, I feel that these may be vendor-specific behaviors? I'm not sure what the "correct" behavior is to can aim for. Do you think these will be disruptive to user's expectations?

The trimming and Safari resizing were thankfully more easy to address. I rebased and pushed 9c5219763c057e9b332564f63a7ad6d4341ca9a1 and ff5d8d4e7d2a9f01922bdc8fb357fd1f44069708 to respectively resolve the issues.

**Comment by nylen** [comment]
Paste behavior and Safari resize behavior look good now.  I couldn't think of an easy way to address the undo stack issues either, so I'll file a separate issue for that.

Also noting that when this change is merged it becomes more important to address the remaining issues with toolbar pinning (#4907) because this adds another situation where the height of editor elements above the toolbar can change.

**Comment by codebykat** [comment]
@aduth Did this fall through the cracks? Looks like it's been marked as ready to merge for awhile and the blocking PR is now resolved.

**Comment by aduth** [comment]
@codebykat : Sorta. I had planned to coordinate this merging with #5445, the other being quite a bit more risky. It doesn't need to, though might benefit from some of the changes there. In any case, I can plan to revive and merge this one this week (looks to not have any merge conflicts surprisingly, but definitely warrants a rebase all the same).

**Comment by codebykat** [comment]
Either way! I was just going through my list and it looked abandoned.

**Comment by aduth** [comment]
Pushed a rebase including a much simpler technique for blocking newlines from pasted text. Because the title is a [controlled input](https://facebook.github.io/react/docs/forms.html#controlled-components), we can replace the newlines in the `onChange` handler, and this will be reflected both in the post edit state and in the next render pass.

See 1213941 for detail. I kept the `preventEnterKeyPress` behavior because while the `onChange` handler would prevent the newlines from enter keypresses, it'd either add a space with the current implementation, or even if we replaced newlines with an empty character, the autosize would already have updated the height in expectation that the newline had been added.

File: client/post-editor/editor-title/style.scss [hunk 5]
```diff
@@ -36,14 +32,17 @@
...
+	min-height: 0;
 	padding-left: 10px; /* Intentionally non-standard to align editor body margin (10px) */
 	padding-right: 10px;
 	border: none;
 	font-family: $serif;
 	font-size: 28px;
 	color: $gray-dark;
 	font-weight: 600;
+	resize: none;
+	-ms-overflow-y: hidden !important;
```
Comment (gwwar) [inline client/post-editor/editor-title/style.scss]: What happens in IE when this rule isn't here?

**Comment by gwwar** [comment]
Found a minor corner case:

double-paste

1. Copy text into clipboard with new lines:
```text
This title
Has
Three Lines
```
2. Paste into title
3. Select Text by double clicking
4. Paste Again

**Comment by gwwar** [comment]
Clicking into the main editor does not deselect any selected title text

**Comment by gwwar** [comment]
I left a few notes on minor corner cases, but this feels pretty solid to me. 👍

File: client/post-editor/editor-title/style.scss (hunk 5 above)
Comment (nylen) [inline client/post-editor/editor-title/style.scss]: An ugly Windows native scrollbar, IIRC.

Comment (aduth) [inline client/post-editor/editor-title/style.scss]: >An ugly Windows native scrollbar, IIRC.

Indeed, this one: https://github.com/Automattic/wp-calypso/pull/4826#issuecomment-212604535

**Comment by aduth** [comment]
>Found a minor corner case:

Ah, this looks to be the same underlying cause as [the issue with Enter I described above](https://github.com/Automattic/wp-calypso/pull/4826#issuecomment-230885559).

>I kept the preventEnterKeyPress behavior because while the onChange replace would prevent the newlines from enter keypresses, it'd either add a space with the current implementation, or even if we replaced newlines with an empty character, the autosize would already have updated the height in expectation that the newline had been added.

I'll plan to poke a bit to see if it's straight-forward to track when the input value could change in this way and whether adding a manual height recalculation could help (and to remove the `onKeyPress` callback).

**Comment by aduth** [comment]
I came up with an alternate solution in 66ad418 to correct the autosize height after replacing if value would have stayed the same. This also removes the need for the `onKeyPress` enter key handling.

Also rebased and squashed a bit, bumped version of `autosize` to latest patch version ([changelog](https://github.com/jackmoore/autosize/blob/master/changelog.md)), and did another test run through all major browsers.

**Pull request merged**